### ISSUE TITLE
Beacon chain loves Trinity

### ISSUE THREAD

**Issue by hwwhww** [ethereum/py-evm#930]
Recently, research team is working on refining [the new spec](https://notes.ethereum.org/s/Syj3QZSxm) and the [**beacon chain** PoC codebase](https://github.com/ethereum/beacon_chain)  (**big shout out to V!**). Basically, there’s a beacon chain that will produce random numbers with RANDAO-based scheme, and we can build Casper logic in the beacon chain to support both beacon chain and shard chains staking.

This issue is for discussing (1) what feature/components are required and (2) what's the blueprint looks like *if we want to implement the beacon chain in the Trinity pyeth world*?

cc @djrtwo , please correct me if I'm wrong. :)

## The certain differences between the main PoW chain and beacon chain
* They both have *Block*, but different fields.
* They both have *State*, but totally different content.
* The main chain uses EVM; beacon chain doesn’t.
* The main chain is still the PoW chain; beacon chain is the full PoS chain (and can implicitly finalize the main PoW chain).

## The potential differences between the main PoW chain and beacon chain
* Beacon chain may use a different serialization scheme instead of RLP.
* Beacon chain may use a different p2p layer for integrating with shard chain networks.
* The main chain has a one-way burns contract; the Casper logic is implemented in beacon chain protocol.
* Main PoW chain, beacon chain, and shards relationship:
     * **Main PoW chain <> beacon**: To add and verify the `main_chain_ref` in beacon block, **the beacon chain validators have to follow the actual canonical main PoW chain**. (see [Beacon-chain-processing](https://notes.ethereum.org/s/Syj3QZSxm#Beacon-chain-processing for more details)
     * **beacon <> shards**: One active validators queue for the validators serve different roles; the validator could be sampled as beacon/shards proposer, notary, attester... etc., so they will have to be both beacon client and shards client.

## Discussion
1. Given those differences, it probably makes sense to have a `beacon` module besides `evm` to put the chain logic?
2. If (1), since there’s no EVM in beacon chain, it will be much simpler here. How many components do we need?
	1. Start by mirroring the reusable `evm` architecture.
        ```
        ├── chains
        ├── constants.py
        ├── db
        ├── exceptions.py
        ├── tools
        ├── utils
        └── validation.py
        ```
	2. Move and reorganize the beacon logics to here.
3. Is there anything we need to be aware of?

**Comment by pipermerriam** [comment]
This all looks generally correct.  

I don't think `beacon` should be a top level module.  I'd suggest either `evm.chains.beacon` or just `evm.beacon` with a completely new `BeaconChain` class that's fully unrelated to the existing `Chain` classes.

**Comment by djrtwo** [comment]
Does this warrant a renaming of the base `evm` module because the beaconchain does not use the EVM.  Maybe `eth`?

**Comment by cburgdorf** [comment]
>Does this warrant a renaming of the base `evm`

Thanks for bringing that up. I was about to throw it in here as well. I'm *very* supportive of the idea to rename the base `evm` module. That should most likely also include the project name itself. Considering that `py-ethereum` is already taken I find `py-eth` with the core module being just `eth` very compelling.

**Comment by carver** [comment]
:+1: for renaming `evm` -> `eth`

Presumably, beacon would go in `eth.chains.beacon`.

**Comment by djrtwo** [comment]
Does `beacon` really belong in `chains`?  Will it be able to inherit from `BaseChain` or `Chain`? If not, it seems strange to place it in there.  The amount of overlap between the EVM protocol and the new beacon/sharding protocol is going to minimal at best. My gut says `eth.evm` and `eth.beacon` as root modules for the separate protocols. That way `eth.beacon.chains` can support `mainnet` and `testnet`s similar to in the current `evm.chains`.

**Comment by pipermerriam** [comment]
My reasoning for putting it in `eth.chains` is that it's called the `BeaconChain` and it is indeed a blockchain, and the `eth.chains` module is where the blockchain abstractions go.  The `BaseChain` will now just be poorly named and should potentially be renamed to `BaseEVMChain` and should potentially get its own namespace, `eth.chains.evm`

**Comment by pipermerriam** [comment]
link: https://github.com/ethereum/py-evm/issues/931

**Comment by hwwhww** [comment]
Thanks for the feedback!

👍  of a more suitable module name.

@pipermerriam @carver 
I can see your point of "`BeaconChain` is indeed a blockchain". I'd point out that there may be not many abstract APIs we can reuse since *maybe* the `BeaconBlock` and `BeaconBlockHeader` are not RLP objects unless we also abstract `BaseBlock`.

BTW, if there's `BaseEVMChain`, that implies there will be a `BaseWASMChain` for eWASM before we decide and start to integrate eWASM into shards?

**Comment by pipermerriam** [comment]
Yes to all of this, and it is completely fine for the `BeaconChain` to have very few methods as well as no RLP.  It's more about logical organization.  

Also, similar to how `EVMChain` defines different VM rules, I'll be pushing for `BeaconChain` to do the same, but using a different name.  This is how we achieve future proof-ness to hard forks in the beacon chain.

**Comment by hwwhww** [comment]
@pipermerriam @djrtwo 

We are thinking about moving [beacon_chain](https://github.com/ethereum/beacon_chain) to Py-EVM. Three parts to build the testnet:
1. Move and refactor the beacon chain repo into `eth`.
2. Implement the python side interface to call [sharding-p2p-poc](https://github.com/ethresearch/sharding-p2p-poc) (golang)
3. Finally, implement the beacon chain server in `trinity`.

* Part 1 and part 2 can start in parallel.
* Please correct me if I am wrong, it seems `trinity` module is updated frequently, but `eth` codebase stable enough and no largely restructure plan, and we can start to in a fork right now?

Another idea is for part 1, we import `py-evm` module like https://github.com/ethereum/beacon_chain/pull/79 to build the useful chain object, and move to Trinity later.

Feedback appreciated. 🙏

**Comment by pipermerriam** [comment]
@hwwhww 

> Please correct me if I am wrong, it seems `trinity` module is updated frequently, but `eth` codebase stable enough and no largely restructure plan, and we can start to in a fork right now?

Yes, `eth` module has been reasonably stable for a few months now.  Biggest change I see coming is finishing adding type hints but that's not going to change APIs.

**Comment by mhchia** [comment]
@pipermerriam 

Currently, we have code to do p2p things in [sharding-p2p-poc](https://github.com/ethresearch/sharding-p2p-poc) written in Golang, while the main logics are in Python. We have communication between Go and Python for both directions. I'm not sure if it makes sense to implement them in the following structures. I really appreciate any feedback on them!

### Transferring data bwteen Python and Go
- Python to Go: when we are to broadcast something, we need to send the data from Python to Go.
- Go to Python: when the p2p layer receive something from other nodes, it needs to verify the data before relaying them, and therefore we need to send the data from Go to Python.

The structure looks like:
![structure](https://i.imgur.com/vF9dNV1.png)
- Python to Go: it can be done through
    - RPC: run the p2p code as a standalone process, setting up a RPC server, and let Python side call.
    - Python bindings: we can do it through `cgo` and dynamic loading. We need to wrap the existing code to a more friendly interface for cgo, compile the code to a shared library, and then load it through `ctypes`. After then we can call the Go functions from Python.
- Go to Python: it can be done through
    - RPC: also set up a RPC server in Python
    - Pipe: set up a pipe for Python and RPC to communicate

### With gRPC for both directions
Currently, for convenience and simpleness, we use RPC([gRPC](https://github.com/grpc/grpc)) to do it for both directions.
![structure_with_both_gRPC](https://i.imgur.com/GU7b01k.png)
It works. However, it might sacrifice the performance, so we might consider replace it with Pipes later.

### With Python bindings from Python to Go
Another implementation for the direction from Python to Go can be done using Python bindings.
![python_bindings](https://i.imgur.com/z6cBaTH.png)
This way we have only one process, and the call from Python to Go becomes a function call to a native library. However, the cons should be
- the interface of the Python bindings must be simple enough. E.g. we should only use raw types in Go in the function signature, instead of any struct.
- harder to debug if there is any issue between Python and Go

Personally, I prefer to stay using RPCs for both directions recently. For future optimization, we can change the "Golang to Python" to Pipe for possible performance issue. I am also willing to change if it makes more sense to use Python bindings from Python to Go, or Pipe from Go to Python, now.

**Comment by pipermerriam** [comment]
I don't think I have sufficient depth to provide a useful opinion.  I have a loose understanding of the approaches you outlined, but I don't know enough details about most any of the exchange mechanisms. :cry: 

General guideline i'd apply is:

1. can you test it?
2. is it hard to debug?
3. is the API or code which uses the API readable?
4. what are the trade-offs?

**Comment by mhchia** [comment]
@pipermerriam 
Thanks a lot for the great guideline! In my current thought, gRPC might be a suitable option, since it is easy to test, to debug, and readable. The tradeoff is the performance, but if it is not a bottleneck, I think it might be tolerable. For the other solutions, I need to dig more into them if we seriously want to replace gRPC with them.

**Comment by hwwhww** [comment]
Hey dear snake charmers!

## tl;dr
Research team is planning to porting [beacon_chain PoC](https://github.com/ethereum/beacon_chain) logic to Py-EVM. @pipermerriam and us had a call this mornining for syncing our understanding. The following notes are our current status and porting plan. It's a very raw outline, many details will be figured out by the future PRs and your feedback will help a lot. (Always appreciated ❤️)

## Beacon chain repo
* Data structure with **Simple Serialize (SSZ)**. For example: [CrystallizedState](https://github.com/ethereum/beacon_chain/blob/master/beacon_chain/state/crystallized_state.py)
* **All** funcations are pure functions.
    - [state_transition.py](https://github.com/ethereum/beacon_chain/blob/db2956c7f883260e728b0febce54d01dcfae2819/beacon_chain/state/state_transition.py)
        - The entrance of state transition is the [compute_state_transition](https://github.com/ethereum/beacon_chain/blob/db2956c7f883260e728b0febce54d01dcfae2819/beacon_chain/state/state_transition.py#L404) function.
    - [helper.py](https://github.com/ethereum/beacon_chain/blob/master/beacon_chain/state/helpers.py) - were defined in [the spec v2.1 - Helper functions](https://notes.ethereum.org/s/Syj3QZSxm#Beacon-chain-state-transition-function) at first, and we implemented them in beacon_chain. Since it's a reference PoC, we tried to keep it similar to the spec.
    - Some proposer logics are in [tests/state/conftest.py - mock_*](https://github.com/ethereum/beacon_chain/blob/db2956c7f883260e728b0febce54d01dcfae2819/tests/state/conftest.py#L276-L390) functions.


## Proposed porting plan
* Add `sharding` or  `beacon_chain` branch.
* Move **SSZ objects (the RLP objects in PoW chain)** 
    * Move them to `eth/ssz/` directory or `eth/beacon_chain/` directory.

* **Implement `Chain` and `DB` classes**
    * Implement `BeaconChain` class in `eth/beacon_chain/chains/beacon_chain_base.py`.
    * Implement `BeaconChainDB` class in `eth/beacon_chain/db/beacon_chain.py`.
    * A demonstration: https://github.com/ethereum/beacon_chain/pull/79/files
* **State execution**
    * We don't need EVM for beacon chain!
    * `eth/vm/forks` is a great design to implment different forks for Ethereum, but for the beacon chain version 1, maybe we can leave the fork abstraction later?
    * Proposal: implement a `eth/beacon_chain/beacon_states.py` (**oh please give a better name if you come up with one**) with some `eth/vm/base.py`-like APIs. e.g., `import_block`, `finalize_block`, `pack_block` ... etc.
* **Implement the attester and proposer logic that can be called from `eth/beacon_chain/beacon_states.py` as well.** 
    ```python
    class BeaconStates:
        block = None  # type: BeaconBlock
        chaindb = None  # type: BeaconChainDB
        crystallized_state = None  # type: CrystallizedState
        active_state = None # type: ActiveState

        def __init__(self, header, chaindb):
            ...

        #
        # State transition APIs
        # 
        ...

        #
        # Attester APIs
        #
        ...

        #
        # Proposer APIs
        #
        ...
    ```
    * Known APIs
        * `make_attestation` - attester
        * `collect_attestations` - proposer
        * `propose_block` - proposer, similar to PoW `mine_block` 
    * **Question: also implement corresponding [builders tool](https://github.com/ethereum/py-evm/blob/master/eth/tools/builder/chain/builders.py) for testing?**
    * TODO: maintain CrystallizedState**s**
   
## Beacon chain important utils -> new packages?
1. [BLS (Boneh–Lynn–Shacham) signature](https://en.wikipedia.org/wiki/Boneh%E2%80%93Lynn%E2%80%93Shacham)
    - Vitalik implemented [BLS in beacon_chain repo](https://github.com/ethereum/beacon_chain/blob/db2956c7f883260e728b0febce54d01dcfae2819/beacon_chain/utils/bls.py) with BN128 curve and will change to BLS12-381 curve.
2. A new serialization protocol - [Simple Serialize (SSZ)](https://github.com/ethereum/beacon_chain/tree/master/ssz)
    * **Question: should we create a `ethereum/pyssz` repo for SSZ or just add a module under Py-EVM for now?**


# P2P plan

## Step-by-step
1. beacon chain Trinity testnet - goal: this year
2. beacon chain cross-client testnet - goal: in 6 months
3. beacon chain + one shard testnet
4. beacon chain + multiple shards testnet

## Alternatives for "beacon chain" testnet
1. With libp2p Python implementation
    - When will be ready to use: ???
        - Part of EF grant program wave 4
    - libp2p is the most promising P2P stack for most Ethereum 2.0 clients.
    - Sharded network needs libp2p + gossipsub.
2. [go-libp2p-daemon](https://github.com/libp2p/go-libp2p-daemon)
    - When will be ready to use: ???
3. 🌟 **[Trinity <--gPRC--> Golang P2P node](https://github.com/ethereum/py-evm/issues/930#issuecomment-420013619) that we're working on** 🌟
    ![](https://camo.githubusercontent.com/8e317aff0a8389a444ecfa4650db091646a9b7f3/68747470733a2f2f692e696d6775722e636f6d2f4755376230316b2e706e67)
    - When will be ready to use:
        - Hopefully, "basic runable network" without attack resistence: 2 months
    - **Action: [Draft](https://notes.ethereum.org/mHOKQ0o_Svymt3_LZNjtTA)**
    - Advantage:
        - The Golang P2P node can be used in other clients too.
        - We can extend it on our own when we need more features.

**Comment by pipermerriam** [comment]
> Question: also implement corresponding builders tool for testing?

Only if you find them useful, probably put them under `eth.tools.builder.beacon` and rename `eth.tools.builder.chain` to something like `eth.tools.builder.evm` to establish clear language about the EVM chain and Beacon chains.

> Question: should we create a ethereum/pyssz repo for SSZ or just add a module under Py-EVM for now?

Probably put it in `eth.ssz` for now and we'll extract to stand-alone library when ready.

**Comment by pipermerriam** [comment]
My question which I surfaced earlier on the call:  Can we build *initial* implementation using DevP2P?  This will save a lot of time/effort/complexity allowing us to re-use the existing p2p code and trinity primitives.

**Comment by hwwhww** [comment]
@pipermerriam We actually considered it as alternatives for beacon chain testnet - No. 4 eariler:

4. Using devp2p for the beacon chain
   - When will be ready to use: soon
   - Disadvantage: it's probably only for Trinity testnet, not a cross-client testnet. We likely have to change it to libp2p eventually.
   - Question: any chance that we can generalize py-evm `p2p` module so that it can be compatible with both devp2p and libp2p?


- I *think* it's quite hard to generalize p2p to be compatible with libp2p since devp2p is a whole system that doesn't have the same functional layers as libp2p. So we give up the idea of generalizing it.
- The tradeoff is that indeed using devp2p will help us get the first Trinity testnet soon in X days! But we still have to overhaul a lot when we move to cross-client testnet in Y days. I'm open to the more incremental plan (using devp2p for the initial implementation). :)
- We still have to allocate researchers to keep working on sharding P2P PoC side since it's essential for sharding network.

Here is my short-term proposal - **two branches**:
1. 2-3 of us port the beacon_chain logic under `/eth/`, and fill all the missing part of the spec.
2. 2-3 of us working on gRPC server "skeleton", just sending random messages on the P2P network. It's useful anyway.

And we will have more knowledge of how complicated it is with gRPC server in 1-2 weeks.

cc @mhchia @djrtwo @ChihChengLiang

**Comment by pipermerriam** [comment]
Cool, then I'd advocate starting with devp2p and only pivoting away when needed.  Saves time and gets a testnet going faster, after which it will likely be easier to swap out the underlying layer, especially if initial build is done with that in mind.

**Comment by carver** [comment]
> Proposal: implement a eth/beacon_chain/beacon_states.py (oh please give a better name if you come up with one) with some eth/vm/base.py-like APIs. e.g., import_block, finalize_block, pack_block ... etc.

My suggestion is to call it a `BeaconVM`, but don't inherit from `BaseVM`, because `BaseVM` is probably too EVM-specific. Continuing to call it a VM is nice because `BeaconChain` probably wants to use it the same way that Eth`Chain` uses the VM classes. For example, to import a block:
- get the block number
- use the number to get the correct VM
- then use the VM to import the block according to the correct rules

I'm speaking with roughly zero familiarity on how `BeaconChain` will work, but it seems likely that we would want to support forks of the beacon chain someday.

**Comment by pipermerriam** [comment]
I'm roughly in the same camp as @carver (with no specific preference yet on actual names for things).  Assuming you feel confident enough that building with the two layer approach `BeaconChain -> BeaconVM` won't slow implementation down too much I think it will be beneficial to build it that way from the start since there will almost definitely be forks in the beacon chain at some point.

**Comment by djrtwo** [comment]
Hesitant to use anything named "VM" in relation to the beacon chain as it does _not_ and will not have a VM. `BeaconStateTransition` or `BeaconST` might be more appropriate.

Agreed that we will fork beacon chain in the future and when we do that, we will either be updating the datastructures and/or the state transition.

**Comment by djrtwo** [comment]
Also, this is a growing [resource about eth2.0](https://notes.ethereum.org/9MMuzWeFTTSg-3Tz_YeiBA)

**Comment by hwwhww** [comment]
@carver @pipermerriam @djrtwo thanks!

* I like the two-layer approach `BeaconChain -> BeaconVM` a lot! `BeaconStates` (or Danny proposed `BeaconStateTransition`, `BeaconStateMachine`) is something similar to `BaseVM` layer but without EVM-state.
* I'm happy to have the hard-forkable abstraction. Also FYI that there's **no block number** in beacon chain, only **time slot number**. Upon receiving a new block, the client has to check if its local clock time is greater than or equal to the minimum timestamp as computed by `GENESIS_TIME + slot_number * SLOT_DURATION`, where `SLOT_DURATION` is 8 seconds. (the number could be adjusted after we have testnet)

**Comment by pipermerriam** [comment]
> Also FYI that there's no block number in beacon chain, only time slot number. 

Does that mean that forks happen at specified *slots* which is roughly equivalent to specified timestamp?

**Comment by hwwhww** [comment]
@pipermerriam I think yes.

And there are **two types of the "beacon chain states" - active state and crystallized state**. The active state recalculation happens when processing every new block; the crystallized state recalculation only happens when the `block.slot_number >= last_state_recalc + CYCLE_LENGTH`, where `CYCLE_LENGTH` is set to 64 slots. In `CrystallizedState`, there are `last_justified_slot` and `last_finalized_slot` fields for Casper FFG finality.

**Comment by djrtwo** [comment]
@pipermerriam yes a fork would happen at a `slot_number` rather than a block height. There is also a notion of a "genesis time". That is slot 0 and slots start counting from there every `SLOT_LENGTH` seconds.

As for the state, there are two sections of the state `Active` and `Crystallized`. They are divided out primarily to reduce the need to frequently hash the large and slower changing `CrystallizedState`. So there are two state roots at any given time -- unless we need one at which point we would do `hash(active_state_root + crystallized_state_root)`

**Comment by pipermerriam** [comment]
Some thoughts that may or may not have already been expressed.

- Lets use `eth.beacon` instead of `eth.beacon_chain` (we'll know what `beacon` signifies), we might even rename `eth.chain` to `eth.evm` or add a namespace and do these as `eth.chains.beacon` and `eth.chains.evm` if we're feeling really ambitious)
- Regardless of the name of the "Forks" (current leader is `BeaconStateMachine`), I'm *lightly* advocating to still use `evm.beacon.forks` since a change to the beacon chain is very likely to still be referred to as a "fork" and that terminology has both general semantic meaning and existing meaning within the code.

Woohoo for the first PR #1322
